projects/btldr/py_client/btldr.py

```diff
@@ -0,0 +1,341 @@
+def flash_time_string(delta):
+    flashed_time = time.localtime(time.time() - delta)
+    return time.strftime("%Y/%m/%d %H:%M:%S", flashed_time)
+
+
+class BtldrManager:
+    def __init__(self):
+        self.db = BtldrDatabase()
+        log = logging.getLogger("root")
+        self.log = log
+        log.info("CAN DBC Initialized")
+
+        # Must be initialized separately
+        self.canbus = None
+
+    """
+    The ping function sends a query message to the target and returns the response.
+
+    Arguments:
+    - ecu_id (int): The ID of the target device to be updated
+    - timeout (int): A timeout (in seconds) to wait when receiving CAN
+        messages
+    """
+
+    def ping(self, ecu_id: int, timeout: int):
+        start = time.time_ns()
+
+        self._send_query(ecu_id)
+        maybe_response = self._receive_query_response(ecu_id, timeout)
+
+        if maybe_response != None:
+            ping_resp = maybe_response
+
+            end = time.time_ns()
+            ping_resp["elapsed_time_ns"] = end - start
+
+            return ping_resp
+        else:
+            return None
+
+    """
+    The flash function is responsible for flashing new firmware onto the target
+    microcontroller. It handles all of the communication with the target.
+
+    Arguments:
+    - ecu_id (int): The ID of the target device to be updated
+    - file (str): A path to the binary that will be uploaded to the target
+    - timeout (int): A timeout (in seconds) to wait when receiving CAN
+        messages
+
+    Returns: A ping response struct of the device after it has been updated
+
+    Raises: Raises an Exception in the event that any stage of the flashing
+        process errored. If this happens, I recommend first retrying, then power
+        cycling the device and retrying, and finally manually updating the device.
+    """
+
+    def flash(self, ecu_id: int, file: str, timeout: int):
+        start = time.time_ns()
+
+        # Reset device into the bootloader
+        logging.info("Resetting target device with ID 0x{:02X}".format(ecu_id))
+        self.software_reset(ecu_id, request_update=True)
+
+        # This is inefficient. Ideally, we should start pinging
+        # immediately over and over again until we get a response, but for the
+        # sake of time and completeness, I decided to hardcode a delay.
+        # Hopefully someone will improve upon this one day.
+        time.sleep(1)  # Wait one second after reset before attempting to ping
+
+        # Query to make sure device is in bootloader
+        ping_resp = self.ping(ecu_id, timeout + 0.5)
+
+        if not ping_resp:
+            logging.critical("Failed to ping device after reset")
+            exit(1)
+            # raise Exception("Failed to ping device")
+
+        if ping_resp["current_image"] != "Updater":
+            logging.critical("Failed to place device in updater")
+            exit(1)
+            # raise Exception("Failed to place device in updater")
```

Comment:
nit: Could delete the commented out exceptions since the logging covers their purpose